text/0000-ptr-to-field.md

```diff
@@ -0,0 +1,395 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This section will go in detail about the semantics and implementation of the types, traits, and methods introduced by this RFC.
+
+## Field types
+
+Field types are just sugar for unit types (`struct Foo;`) that are declared right next to the `Parent` type.
+
+Like so,
+```rust
+struct Person {
+    pub name: String,
+    pub(super) age: u32,
+    id: u32,
+}
+
+pub struct Person.name;
+pub(super) struct Person.age;
+struct Person.id;
+```
+
+All field types will implement the following traits: `Debug`, `Clone`, `Copy`, `PartialEq`, `Eq`, `PartialOrd`, `Ord`, `Hash`, with the same implementation for unit types.
+
+They will also need to interact with the privacy rules for fields, and will have the same privacy as the field that they are derived from.
+
+You can make a field type for the following types of types (some example syntax is shown)
+* tuples `(T, U, ...)`
+    * `<(T, U)>.0`
+* tuple structs `Foo(T, U, ...)`
+    * `Foo.0`
+* structs `struct Foo { field: Field, ... }`
+    * `Foo.field`
+* unions `union Foo { field: Field }`
+    * same syntax as structs
+    * constructing a field type is `unsafe`, this is because accessing fields of `union`s is unsafe
+
+The compiler can decide whether to actual generate a field type, this is to help compile times (if you don't use field types, then the compiler shouldn't slow down too much because of this feature).
+
+## `*const T`/`*mut T`
+
+`project_unchecked` and `wrapping_project` will both live inside of `core::ptr`.
+
+We need both `project_unchecked` and `wrapping_project` because there are some important optimization available inside of LLVM related to aliasing and escape analysis. In particular the LLVM `inbounds` assertion tells LLVM that a pointer offset stays within the same allocation and if the pointer is invalid or the offset does not stay within the same allocation it is considered UB. This behaviour is exposed via `project_unchecked`. This can be used by implementations of the `Project` trait for smart pointers that are always valid, like `&T` to enable better codegen. `wrapping_project` on the other hand will not assert `inbounds`, and will just wrap around if the pointer offset is larger than `usize::max_value()`. This safe defined behaviour, even if it is almost always a bug, unlike `project_unchecked` which is UB on invalid pointers or offsets.
+
+This corresponds with `core::ptr::add` and `core::ptr::wrapping_add` in safety and behaviour. `project_unchecked` and `project` need to be implemented as intrinsics because there is no way to assert that the pointer metadata for fat pointers of `Field::Parent` and `Field::Type` will always match in general without some other compiler support. This is necessary to allow unsized types to be used transparently with this scheme.
```

Comment:
So wait, how does this interact with unsized types exactly? This needs to be spelled out.

Since this is specifically to give vocabulary to the direct manipulation of the pointers, I suspect that this should require that the field pointer's metadata is derivable from the parent metadata. Note that this doesn't require equivalent metadata: a theoretical pointer with two metadatas could be split into two child fat pointers with one or the other metadata.

The intrinsics live in [`core::intrinsics`](https://doc.rust-lang.org/core/intrinsics/index.html), of course.

Reply:
> Note that this doesn't require equivalent metadata: a theoretical pointer with two metadatas could be split into two child fat pointers with one or the other metadata.

Actually, you can only have 1 unsized field on a type, and it must be the last field, so the metadata will be exactly the same.

> The intrinsics live in core::intrinsics, of course.

Yes, I will make it clear where everything lives in the next revision. I see now that it wasn't clear enough